### PR TITLE
Prevent duplicate spectra in Indirect F(Q)Fit Multiple Table

### LINKED ISSUE #29244: Bug for adding multiple copies of a spectra in indirect fq fit
In indiect data analysis on the f(q) fit on the multiple spectra option you can add multiple copies of a single spectra to the table. if you then switch to a different workspace and back the duplicates are removed. there should be a check to prevent duplicated being added.

### PR DESCRIPTION
**Pull request by sf1919**
**Description of work.**
Code altered to prevent spectra duplicates in the F(Q)Fit Multiple Spectra Table. Removed obsolete methods and tests.

**To test:**

Test 1
1) Open Indirect Data Analysis
2) go to the F(Q) fit tab and switch to multiple input mode.
3) click add workspace, browse to a suitable file and open
4) Add different Width and EISF parameters to the table. Try adding ones that have already been added - should not duplicate them.

Test 2
1) Open Indirect Data Analysis
2) go to the F(Q) fit tab and stay on the single input tab
3) click on browse, find a suitable file and open
4) Check that the Fit Parameter and Width/EISF drop downs work.

Fixes #29244  

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/scientific_interfaces/Indirect/FqFitDataPresenter.cpp [hunk 1]
```diff
@@ -36,8 +36,11 @@ FqFitDataPresenter::FqFitDataPresenter(FqFitModel *model, IIndirectFitDataView *
 }
 
 void FqFitDataPresenter::handleSampleLoaded(const QString &workspaceName) {
+  const auto workspaceName1 = workspaceName.toStdString();
+  auto workspace = Mantid::API::AnalysisDataService::Instance().retrieveWS<MatrixWorkspace>(workspaceName1);
```
Comment (Harrietbrown) [inline qt/scientific_interfaces/Indirect/FqFitDataPresenter.cpp:40]:
```suggestion
  auto workspace = 
      Mantid::API::AnalysisDataService::Instance().retrieveWS<MatrixWorkspace>(workspaceName.toStdString());
```

File: qt/scientific_interfaces/Indirect/FqFitDataPresenter.cpp [hunk 2]
```diff
@@ -124,42 +127,47 @@ void FqFitDataPresenter::handleParameterTypeChanged(const QString &parameter) {
 }
 
 void FqFitDataPresenter::setDialogParameterNames(FqFitAddWorkspaceDialog *dialog, const std::string &workspace) {
```
Comment (Harrietbrown) [inline qt/scientific_interfaces/Indirect/FqFitDataPresenter.cpp:129]:
```suggestion
void FqFitDataPresenter::setDialogParameterNames(FqFitAddWorkspaceDialog *dialog, const std::string &workspaceName) {
```

File: qt/scientific_interfaces/Indirect/FqFitDataPresenter.cpp [hunk 3]
```diff
@@ -124,42 +127,47 @@ void FqFitDataPresenter::handleParameterTypeChanged(const QString &parameter) {
...
+  FqFitParameters parameters;
   try {
-    addWorkspace(m_fqFitModel, workspace);
+    auto workspace1 = Mantid::API::AnalysisDataService::Instance().retrieveWS<MatrixWorkspace>(workspace);
+    parameters = m_fqFitModel->createFqFitParameters(workspace1.get());
```
Comment (Harrietbrown) [inline qt/scientific_interfaces/Indirect/FqFitDataPresenter.cpp:133]:
```suggestion
    auto workspace = Mantid::API::AnalysisDataService::Instance().retrieveWS<MatrixWorkspace>(workspaceName);
    parameters = m_fqFitModel->createFqFitParameters(workspace.get());
```

File: qt/scientific_interfaces/Indirect/FqFitDataPresenter.h [hunk 4]
```diff
@@ -30,9 +30,12 @@ class MANTIDQT_INDIRECT_DLL FqFitDataPresenter : public IndirectFitDataPresenter
 private slots:
   void hideParameterComboBoxes();
   void showParameterComboBoxes();
+  /* void updateAvailableParameters(FqFitParameters &parameters);
+   void updateAvailableParameters(const QString &type, FqFitParameters &parameters);*/
```
Comment (Harrietbrown) [inline qt/scientific_interfaces/Indirect/FqFitDataPresenter.h:34]:
```suggestion
```

File: qt/scientific_interfaces/Indirect/FqFitDataPresenter.h [hunk 5]
```diff
@@ -30,9 +30,12 @@ class MANTIDQT_INDIRECT_DLL FqFitDataPresenter : public IndirectFitDataPresenter
...
   void updateAvailableParameters();
-  void updateAvailableParameterTypes();
   void updateAvailableParameters(const QString &type);
+  // void updateAvailableParameterTypes();
```
Comment (Harrietbrown) [inline qt/scientific_interfaces/Indirect/FqFitDataPresenter.h:37]:
```suggestion
```

File: qt/scientific_interfaces/Indirect/FqFitDataPresenter.h [hunk 6]
```diff
@@ -41,6 +44,7 @@ private slots:
   void updateActiveDataIndex();
   void updateActiveDataIndex(int index);
   void setSingleModelSpectrum(int index);
+  /* void handleParameterTypeChanged(const QString &parameter, FqFitParameters &parameter1);*/
```
Comment (Harrietbrown) [inline qt/scientific_interfaces/Indirect/FqFitDataPresenter.h:47]:
```suggestion
```

File: qt/scientific_interfaces/Indirect/test/FqFitModelTest.h [hunk 7]
```diff
@@ -130,15 +98,6 @@ class FqFitModelTest : public CxxTest::TestSuite {
     TS_ASSERT_EQUALS(m_model->getWidths(TableDatasetIndex{0})[1], "f1.FWHM");
   }
 
-  void test_that_getEISF_will_return_an_empty_vector_if_there_are_no_EISFs() {
```
Comment (Harrietbrown) [inline qt/scientific_interfaces/Indirect/test/FqFitModelTest.h:133]: This should still be left tyhere. and can you write an equivilant test for getWidths

**Review by Harrietbrown — CHANGES_REQUESTED** [review]

File: qt/scientific_interfaces/Indirect/test/FqFitModelTest.h [hunk 8]
```diff
@@ -91,13 +93,32 @@ class FqFitModelTest : public CxxTest::TestSuite {
         "f1.FWHM");
   }
 
+  void test_that_zeroWidths_returns_true_if_the_workspace_contains_no_widths() {
```
Comment (Harrietbrown) [inline qt/scientific_interfaces/Indirect/test/FqFitModelTest.h:96]: Why can this not be the same format as its EISF equivilant?

Comment (sf1919) [reply qt/scientific_interfaces/Indirect/test/FqFitModelTest.h:96]: I tried that (see commit 2467fe9) but the test failed. I went back to the original code and copied the original test that tested getWidths but removed the zeroWidths assertion as that method no longer exists.

Comment (Harrietbrown) [reply qt/scientific_interfaces/Indirect/test/FqFitModelTest.h:96]: Ok so long as it is not a reversion

**Review by Harrietbrown — APPROVED** [review]
Good job, I know it took a while but well done getting this working. Duplicates are not longer possible in FqFit as described.

**Review by DavidFair — APPROVED** [review]
LGTM, I've resisted the temptation to ask for moving `Mantid::API::AnalysisDataService::Instance()` into an anonymous namespace, as it's creeping into refactoring.

Next time you look at this file could you try and consolidate all the refs and remove the boilerplate into one place like:
```c++
namespace{
auto &ads_instance = Mantid::API....::Instance()
}

FooClass::SomeMethod(...){
   ads_instance.retrieveWS<MatrixWorkspace>(wsName);
}
```

**Comment by sf1919** [comment]
> 
> 
> LGTM, I've resisted the temptation to ask for moving `Mantid::API::AnalysisDataService::Instance()` into an anonymous namespace, as it's creeping into refactoring.
> 
> Next time you look at this file could you try and consolidate all the refs and remove the boilerplate into one place like:
> 
> ```c++
> namespace{
> auto &ads_instance = Mantid::API....::Instance()
> }
> 
> FooClass::SomeMethod(...){
>    ads_instance.retrieveWS<MatrixWorkspace>(wsName);
> }
> ```

@Harrietbrown - is this something that can be incorporated into the issue you're working on to remove the single input for F(q)Fit?

**Comment by Harrietbrown** [comment]
Yep wont take more than a second

**Pull request merged**